### PR TITLE
Fix preventing some effects from working

### PR DESCRIPTION
**Pull request by chrisgonzales**
Missing break from switch case for getting pixel color.

### CODE REVIEW

**Comment by FeatherKing** [comment]
I confirmed that this fixed 'Peek' and all of the patterns that were crashing my esp32

**Comment by intermittech** [comment]
I can confirm, this fixes the crashes on certain effects!

**Review by illuxions — APPROVED** [review]
this also fixed the peek and certain effects issues after today's merge.

**Comment by lonestriker** [comment]
This fixes my crashes as well, thanks for the PR.  Previously, saving updates to presets crashed WLED consistently.

I'm having other issues though with the multi-pin mod though.  I'm trying to run an 800 pixel megatree using 4 outputs with 200 LED each.  I get random flashing on some of the output pins where the most of the string flashes a different color and then back to the correct effect.  Basically like something's simultaneously updating the same data output as my LED pin.  I can't seem to pinpoint the issue though, it doesn't seem to follow the particular pin I use if I shuffle them around.

Has anyone successfully run a large number of LEDs across 4+ outputs before?

**Comment by illuxions** [comment]
> This fixes my crashes as well, thanks for the PR. Previously, saving updates to presets crashed WLED consistently.
> 
> I'm having other issues though with the multi-pin mod though. I'm trying to run an 800 pixel megatree using 4 outputs with 200 LED each. I get random flashing on some of the output pins where the most of the string flashes a different color and then back to the correct effect. Basically like something's simultaneously updating the same data output as my LED pin. I can't seem to pinpoint the issue though, it doesn't seem to follow the particular pin I use if I shuffle them around.
> 
> Has anyone successfully run a large number of LEDs across 4+ outputs before?

Lone,

Are you trying to run directly from the outputs of the ESP without a buffer or level shifter? You need the beginning of each string to be inches from the output of the ESP as the output is 3.3v, there's a regulator that drops the 5v input to 3.3 and that's what it runs on and outputs.  3.3V isn't quite enough and can cause all sorts of problems with any distance.  The Dig quad uses a buffer, I prefer to use hex inverting schmitt triggers because I have a few of them from a previous company I worked for(many rails) and using a pair is an excellent noise decoupler.   

I'm running 8 outputs from an ESP32 with 100,100,100,100,100,150,200,150 https://i.imgur.com/8U1wUdt.jpg 100 for each of the outer palms, 150 on the inner, 200 on the front eave(about 15' from the controller to first pixel with zero trouble), and 150 to C9 pixels that I WAY overpaid for on the back bushes.  The windows are unique controllers because they are inside.  Make sure NUM_STRIPS is correct as well as each STRIP#_LEDCOUNT, then you need to configure the segments if you want them to operate independently.   The biggest thing is distance to first pixel or using something to gain up the output of the ESP.  Here's my breadboard with 6 channels out wired, 2 more ready https://i.imgur.com/0jSGhTa.jpg

**Comment by lonestriker** [comment]
@illuxions thanks much for the reply.  Glad to hear that things are working well and it's likely something with my setup and nothing inherent with the fork itself.  I've run the strings individually with stock WLED without the flashing issue, but timings between the two versions might be different enough that would cause issues.

I've previously run other controllers with stock WLED without logic level shifters and they seemed to be fine.  I am not running with logic shifters on my ESP32-POE controller, but that was going to be my next debugging step.  Thanks much for the suggestion.

I would love to order the dig quad if it could support some of the new cheap ethernet-enabled ESP32 boards.  Wi-Fi generally works OK for my xLights sequences, but it lags or hiccups often enough to be a problem.

**Comment by lonestriker** [comment]
@illuxions Logic level shifters did the trick! No more random flashes. So, it appears that my ESP32 + WS2811 LEDs are more noisy or sensitive to 3.3 V => 5V using a single output vs. the multi-pin mod.  It's a rat's nest of wires though for 4 pins let alone my intention to use 6-8 pins.  The digi quad would come in very handy here if it supported the cheap ESP32 ethernet boards that are now becoming available.

The multi-pin mod would make the digi quad very competitive with the other commercial E1.31 controllers for xLights if it only had an ethernet port.

Thanks again.

**Comment by illuxions** [comment]
@lonestriker glad it worked out.  I had the glitchy issues with 4 controllers beside each other and rummaged through my office to find the schmitt triggers then all of the problems instantly went away.  I like that ethernet ESP32, but ouch it's pricey.  I like the self contained effects of WLED but had planned to upgrade some Dan Kulp toys I bought earlier this year https://i.imgur.com/qGp54AE.jpg , though I ended up being too lazy.  Maybe next year.

**Comment by lonestriker** [comment]
@illuxions Supposedly, these cheap $6 [ESP32-WT32 ethernet boards](https://www.seeedstudio.com/Ethernet-module-based-on-ESP32-series-WT32-ETH01-p-4736.html) work.  They're just missing the USB connector so you'll have to use an FTDI serial board to program them the first time.  Then flash OTA with WLED afterward.  They're not much more than a standard ESP32, but you'll have to wait for shipment from China.  I should be taking delivery next week hopefully to test.

The ESP32-POE boards are pricey, running around $25 and those work flawlessly with WLED.  I'm hoping these new boards work just as well as they would be the perfect boards to pair with a digi quad given how cheap they are. Would need to jumper the pins though as they're not the same small form factor.

**Comment by lonestriker** [comment]
> @lonestriker glad it worked out. I had the glitchy issues with 4 controllers beside each other and rummaged through my office to find the schmitt triggers then all of the problems instantly went away. I like that ethernet ESP32, but ouch it's pricey. I like the self contained effects of WLED but had planned to upgrade some Dan Kulp toys I bought earlier this year https://i.imgur.com/qGp54AE.jpg , though I ended up being too lazy. Maybe next year.

From one procrastinator to another I can relate;  this was finally the year I made my megatree after a couple years of just playing and planning.  I have a few of those Falcon smart receiver boards as well :).  I don't have any of the Beaglebone devices though.  I just use either the Falcon 16V3 or ESP8266/ESP32 boards with WLED.  I'm starting to outgrow the wireless-only boards though since I can't seem to get smooth xLights performance any longer (even trying different, dedicated mesh networks for the show devices.)

**Comment by intermittech** [comment]
Ethernet for my boards ([Dig-Uno](https://quinled.info/2018/09/15/quinled-dig-uno/) and [Dig-Quad](https://quinled.info/2020/06/17/quinled-dig-quad/)) is currently in development, I hope to introduce it somewhere Q1 2021 but it's a complex little work. But in theory once it's done you can just pull the ESP8266/ESP32 off and plop the QuinLED-ESP32-E(thernet) (working name) on there, done! ;)

I'm also going to try and make it cost effective!

But yeah, lots to do, verify, prototype, test, etc.. before I can release that. :)

**Comment by lonestriker** [comment]
If you and Aircookie aren't careful, you might end up creating a low-cost, feature-rich software + hardware competitor to the satellite Falcon and Beaglebones pixel controllers :) At least in the E1.31 space where we don't need any of the older serial protocols, using an off-the-shelf ESP32 opens up lots of possibilities.

**Comment by DeviousPenguin** [comment]
Tested this fix for my ESP8266 based setup using 2 pins, and this seemed to the effects and 'peek' now works again too

**Comment by lonestriker** [comment]
Just posted this on Aircookie's WLED github, copying here for others to see.  I made the changes to this multipin mod version and things mostly work as expected.  I've got a bit more testing to do though, but so far, so good.  The board seems more stable than my ESP32-POE board in terms of flickering without using a logic-level shifter.

=====================================

OK, so I got three of the boards delivered today

TL;DR: they work!  But with some code changes needed.

Some notes:

* I couldn't figure out a way to flash using the existing firmware (I didn't really try too hard other than looking for a Wi-Fi AP or plugging into ethernet and hoping for a new DHCP address showing up on my network.)
* I used a standard FTDI adapter to flash new firmware on these boards.  You have to jumper GPIO-0 (zero) with ground to program since there's no button to do that on these devices.
* Use TX0 and RX0 pins and **NOT** TXD and RXD ones to connect to FTDI.
* I used 5V on my FTDI to program and then run the board to watch logs (3.3V isn't enough to power this WT32-ETH01 board and I got brownout errors.)
* I verified that Tasmota works with it compiling a version with USE_ETHERNET defined and uncommenting out the my_user_config.h ETH_TYPE, ETH_ADDR and ETH_CLKMODE settings for the WT32-ETH01
* You also have to configure the [WT32-ETH01 Tasmota template](https://tasmota.github.io/docs/ESP32/#wt32-eth01) as well before the ethernet adapter works.
* I was able to take the Tasmota settings and hack them into WLED and can use the WT32-ETH01 just the same as the ESP32-POE adapter (minus of course the POE power injection.)
* I'm able to use this board + modified WLED to run as part of my E1.31 Christmas lights display.

So definitely grab some of these boards if you need ethernet for your WLED runs.  Particularly useful for lag-free E1.31 xLights displays.  I'll clean up the code and submit a PR after some more testing.

**Comment by jeffeb3** [comment]
Did I miss something? Why hasn't this been merged? Seems like a slam dunk fix.

Edit: Ok, upstream changes (aircookie's version) seem to have gobbled up the need for this fork. And it doesn't look like the aircookie version needs this fix.

**Comment by fusionstream** [comment]
Line numbers no longer match with the pull request but there seems to be 2 now at 620 (+ the end) and 802 (+ the end) - you need a total of 4 more _break;_ s

**Comment by peacepenguin** [comment]
sorry, i totally missed this PR! ill take a look at merging it soon!

**Comment by peacepenguin** [comment]
Thanks @chrisgonzales this fix works perfectly to fix the "peek" issue I was able to reproduce prior to merging your fix!

@fusionstream was there another issue you ran into that isn't addressed by the Pull request from chrisgonzales? let me know! thanks!

**Comment by fusionstream** [comment]
> 
> Line numbers no longer match with the pull request but there seems to be 2 now at 620 (+ the end) and 802 (+ the end) - you need a total of 4 more _break;_ s
@peacepenguin not sure if the line numbers are accurate anymore as the pr only adds 2 more breaks;

The other things was just adding support for the Wemos Lolin32 board. It's really just a board definition.

`[env:lolin32]`
`board = lolin32`
`platform = espressif32@2.1.0`
`build_unflags = ${common.build_unflags}`
`build_flags = ${common.build_flags_esp32}`
`lib_ignore =`
`  ESPAsyncTCP`
`  ESPAsyncUDP`
`upload_protocol = esptool `
